evm/utils/test_builder/normalization.py

```diff
@@ -0,0 +1,103 @@
+from collections.abc import (
+    Iterable,
+    Mapping,
+)
+
+from cytoolz import (
+    assoc_in,
+    compose,
+    identity,
+)
+from eth_utils import (
+    big_endian_to_int,
+    decode_hex,
+    is_0x_prefixed,
+    is_bytes,
+    is_hex,
+    to_canonical_address,
+)
+import cytoolz.curried
+import eth_utils.curried
+
+
+def state_list_to_dict(state_list):
+    """Convert state definitions in list form to the canonical dict form.
+
+    Example of state list elements:
+
+        - `(address, "balance", 1)`
+        - `(address, "nonce", 2)`
+        - `(address, "code", b"3")`
+        - `(address, "storage", 4, 5)`
+
+    For storage, the second to last entry specifies the slot and the last one the value.
+    """
+    d = {}
+    for state_item in state_list:
+        d = assoc_in(d, state_item[:-1], state_item[-1])
```

Comment:
*functional nonsense*

```python
return cytoolz.merge(*(
    assoc_in({}, state_item[:-1], state_item[-1])
    for state_item
    in state_list
))
```

Reply:
I like it, that's nicer (but needs `deep_merge` instead of `merge`). With 
```
deep_merge(*(
    assoc_in({}, state_item[:-1], state_item[-1]) if isinstance(state_item, Mapping) else state_item
    for state_item
    in state_list
))
```
this would be even more flexible (then one can switch from the list to dict definition at any level:
```
[
    ("0xaabb", "balance", 3),
    ("0xaabb", "storage", {
        4: 5,
    }),
    {"0xbbcc", {
        "balance": 6,
        "nonce": 7
    }
]
```